Include attachments and featured images for posts being exported

nate-allen opened the pull request:
I closed my previous PR on this issue because it was stuck on a failed PHP 5.6 test that doesn't apply anymore.

This pull request fixes an issue where attachments are not included when exporting posts, unless you choose the "all" option.

A unit test is also included.

Trac ticket: [17379](https://core.trac.wordpress.org/ticket/17379)

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

nate-allen committed:
Get attachment IDs for each post being exported

Currently, attachmets are only exported if you choose to export
with the 'all' option. This change will get the attachment IDs
of all of the posts being exported when the 'all' option isn't
used.

nate-allen committed:
Add unit test to ensure that attachments are included

nate-allen committed:
Add newline at end of file

mukeshpanchal27 commented on src/wp-admin/includes/export.php:
```diff
@@ -144,6 +144,40 @@ function export_wp( $args = array() ) {
+			$posts_in     = array_map( 'absint', $chunk );
+			$placeholders = array_fill( 0, count( $posts_in ), '%d' );
+
+			// Prepare the SQL statement for attachment ids
```
```suggestion
			// Prepare the SQL statement for attachment ids.
```

petitphp commented on src/wp-admin/includes/export.php:
```diff
@@ -144,6 +144,40 @@ function export_wp( $args = array() ) {
 	$post_ids = $wpdb->get_col( "SELECT ID FROM {$wpdb->posts} $join WHERE $where" );
 
+	// Get IDs for the attachments of each post, unless all content is already being exported.
+	if ( 'all' !== $args['content'] ) {
```
Those extra queries will be executed when exporting attachments, where it is not useful.

We could update the condition to skip this part if the `$args['content']` is set to `all` or `attachment`. I think it would be safe to hard-code those two values, but we could also consider adding a filter to allow developers to include custom post types to the list.

hellofromtonya replied:
>We could update the condition to skip this part if the $args['content'] is set to all or attachment.

+1

hellofromtonya replied:
>could also consider adding a filter to allow developers to include custom post types to the list.

Yes, a filter could be added for customization. That said, adding a filter would need to wait until 6.6, as it's too late in the 6.5 cycle given adding a filter is considered an `enhancement` and 6.5 is now in beta.

nate-allen replied:
I updated the condition to skip if the export is `all` or `attachment`. I did it in a way that will make it easy to add a filter in the future. We would just filter the `array( 'all', 'attachment' )` array.

hellofromtonya commented:
>Request: Have Tools>Export "specific author" generate the same metadata as the "all content" export, so that Export "specific author" will also include metadata reference to the Featured Image.

The Trac ticket is focused on exporting for a _"specific author"_ rather than _all authors_. The PR works for both. Is this intentional to include when no author is selected (i.e. to do all authors)?

If yes it should also work for all authors, what might be the risks or impacts on sites with a large or massive amounts of content?

nate-allen commented:
> The Trac ticket is focused on exporting for a _"specific author"_ rather than _all authors_. The PR works for both. Is this intentional to include when no author is selected (i.e. to do all authors)?

The Trac ticket mentions the bug happens when you filter by specific author, but it actually happens when you do any filter other than "All content"

This patch fixes that by getting all the attachments for the posts being exported if `$args['content']` isn't `all`. I'll update the patch to not do this if the posts being exported are `attachment`

> If yes it should also work for all authors, what might be the risks or impacts on sites with a large or massive amounts of content?

The patch takes this into consideration by splitting the work into [chunks of 20 post IDs](https://github.com/WordPress/wordpress-develop/pull/5720/files#diff-23fb643f7d0c9700e840c2fa179744f825c1dc6dc22a975a660f87a05c3ef061R149) at a time.

nate-allen committed:
Update condition to skip if export is 'all' or 'attachment'

hellofromtonya commented on src/wp-admin/includes/export.php:
```diff
@@ -144,6 +144,40 @@ function export_wp( $args = array() ) {
 
+	// Get IDs for the attachments of each post, unless all content is already being exported.
+	if ( ! in_array( $args['content'], array( 'all', 'attachment' ), true ) ) {
+		foreach ( array_chunk( $post_ids, 20 ) as $chunk ) {
```
There's already a [chunking approach in the same file](https://github.com/WordPress/wordpress-develop/blob/6.4/src/wp-admin/includes/export.php#L544):

```
// Fetch 20 posts at a time rather than loading the entire table into memory.
while ( $next_posts = array_splice( $post_ids, 0, 20 ) ) {
```

Could that approach also be used here for consistency? 

Are there benefits to this approach over the other?

nate-allen replied:
The approach you mentioned actually is similar, but it modifies the array instead of creating new ones. I went ahead and used it for consistency but I also needed to create a copy of the `$post_ids` so it wouldn't get modified.

hellofromtonya commented on src/wp-admin/includes/export.php:
```diff
@@ -144,6 +144,40 @@ function export_wp( $args = array() ) {
+					"
+				SELECT meta_value
+				FROM {$wpdb->postmeta}
+				WHERE {$wpdb->postmeta}.post_id IN (" . implode( ',', $placeholders ) . ")
```
Micro-optimization: the `implode()` can be moved up and assigned to a variable to be reused in both instances.

nate-allen replied:
Okay, I assigned it to a variable for reuse 👍

hellofromtonya commented:
>The Trac ticket mentions the bug happens when you filter by specific author, but it actually happens when you do any filter other than "All content"

Aha thanks @nate-allen for clarifying. Makes sense 👍

hellofromtonya commented on tests/phpunit/tests/export/attachments.php:
```diff
@@ -0,0 +1,66 @@
+	 * @runInSeparateProcess
+	 * @preserveGlobalState disabled
+	 */
+	public function test_export_includes_attachments_for_specific_author() {
```
It would be great to have more test coverage to cover the other selectable states such as:

* Selecting to export all posts but not selecting a specific author.
* Selecting to export all pages but not selecting a specific author.
* Selecting to export all pages from a specific author.

nate-allen replied:
We now have the following unit tests:

1. When exporting all content, all posts and pages are exported as well as their attachments
2. When exporting posts, all posts are exported along with their attachments
3. When exporting pages, all pages are exported along with their attachments
4. When exporting pages by a specific author, their pages and attachments are exported
5. When exporting posts by a specific author, their posts and attachments are exported

hellofromtonya commented:
It's really close for commit. Left some comments to finalize it.

nate-allen committed:
Use while loop and array_splice to chunk the data

This is more consistent to how it is being done elsewhere in the
file and is a bit more memory efficient as well.

nate-allen committed:
Create placeholders to be used in multiple places

nate-allen committed:
Create a copy of post_ids to avoid modifying the original

nate-allen committed:
Add additional unit tests

nate-allen committed:
Each array value should start on a new line

mukeshpanchal27 commented on src/wp-admin/includes/export.php:
```diff
@@ -144,6 +144,52 @@ function export_wp( $args = array() ) {
 
+	// Get IDs for the attachments of each post, unless all content is already being exported.
+	if ( ! in_array( $args['content'], array( 'all', 'attachment' ), true ) ) {
+		// Array to hold all additional IDs (attachments and thumbnails)
```
```suggestion
		// Array to hold all additional IDs (attachments and thumbnails).
```

nate-allen committed:
Add full stop at end of comments

petitphp commented on src/wp-admin/includes/export.php:
```diff
@@ -144,6 +144,52 @@ function export_wp( $args = array() ) {
+					"
+				SELECT meta_value
+				FROM {$wpdb->postmeta}
+				WHERE {$wpdb->postmeta}.post_id IN ($in_placeholder)
+				AND {$wpdb->postmeta}.meta_key = '_thumbnail_id'
```
Small nitpick, we could drop the `{$wpdb->postmeta}.` in the conditions since we're only using a single table in the SQL query. This would match the format of the query above.

hellofromtonya replied:
The `{}` curly braces could be dropped, but it's currently following the predominate pattern in this file.

That said, there's a mixture of with and without in the export file and throughout Core. So either approach is fine.

>This would match the format of the query above.

Agreed. The structure of both queries should match for consistency and readability.

petitphp reviewed (approved):
Thanks for all the work on this. This looks ready for me.

I've left one small nitpick, but happy to see it merge as is.

hellofromtonya commented:
Thank you @nate-allen and everyone for the follow-ups to code review and suggestions. Reviewing today for commit.

hellofromtonya commented on tests/phpunit/tests/export/attachments.php:
```diff
@@ -0,0 +1,216 @@
+ * @group attachments
+ * @ticket 17379
+ */
+class Test_Export_Includes_Attachments extends WP_UnitTestCase {
```
This test class covers the attachment part of `::export_wp`. But the test location, file name, and test class name need to change. Below, I'll share context, information, and handbook links to help explain the changes.

Note: I'll push a commit shortly to take care of these changes.

----------------------------

Sharing the longer details:

The test suite coding standards are [located here in the handbook](https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/).

What is the function under test?
`export_wp()`

This means the test class is for `export_wp()` coverage. Specifically for this bug, the coverage is for only the attachment fix within the `export_wp()`.

Where should the file be located?
`tests/phpunit/tests/admin/` directory as part of the `admin` group of tests. 

The `export_wp()` function lives in the `/admin/includes/export.php` file as part of the export group of admin functionality.

Notice the test files in the files and functions in `/admin/includes/` are located in `tests/admin/`. 

What should the filename be?
`exportWp.php`.

Notice in the [Test Classes](https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#test-classes) standard, the test filename is the function under test's name in camelCase.

What should the test class name be?
`Tests_Admin_ExportWp` per the [same standard](https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#test-classes), which defines the naming as
```
Tests_[TheGroup]_FunctionNameUnderTest
```

where the `TheGroup` is the root or primary group for the set of tests, i.e. identified by subdirectory which is in this case is the `admin` group of tests.

hellofromtonya replied:
Done in https://github.com/WordPress/wordpress-develop/pull/5720/commits/6c9d28d204579317db9a9c1614bb31a67a589452 ✅

hellofromtonya committed:
Test improvements.

Changes the test class to cover the export_wp() function, i.e. the function under test.

1. Relocates to tests/admin/.
2. Renames the test file to function's name in camelCase.
3. Renames the test class to follow Tests_[GroupName]_FunctionName.
4. Changes the groups to admin and export.
5. Renames tests to test_should_[what_it_should_do] format.
6. Moves the ticket annotation to the test methods.
7. Adds covers annotation.
8. Adds test failure reason when there are multiple assertions (i.e. to help identify which assertion failed).
9. Uses assertSame() as part of an effort in Core to reduce assertEquals() usage.
10. Switch to dataProvider.
11. Update docblocks.

hellofromtonya commented:
To ensure there are no merge conflicts, a rebase is needed against the current WordPress Develop `trunk`. Will force push shortly.

Update: Done ✅

hellofromtonya committed:
Remove {} around $wpdb->postmeta

hellofromtonya reviewed (approved):
* Confirmed tests fail without the changes in `export_wp()` ✅
* Confirmed this PR resolves the issue for the featured images. [See test report](https://core.trac.wordpress.org/ticket/17379#comment:57) ✅
* Tests are per Core's test coding standards and practices ✅
* The changes in `export_wp()` are using the same code design as preexisting in the `export.php` file ✅

This patch is ready for commit ✅

hellofromtonya commented:
Committed via https://core.trac.wordpress.org/changeset/57681.
